finality-aleph/src/block/substrate/verification/cache.rs

```diff
@@ -403,6 +403,16 @@ where
             maybe_equivocation_proof,
         })
     }
+
+    fn own_block(&self, header: &Header) -> bool {
+        // on average: ~2000 elements in the cache, one check per second
+        for (cached_header, certainly_own) in self.equivocation_cache.values() {
```

Comment:
Come on, don't loop, add some indexing which makes this check O(1).

Reply:
Agree, I guess one way to do that is to convert `equivocation_cache` into `HashMap<Slot, H>` and also keep `certainly_own: HashSet<_>` where you insert `header.hash()` of each certainly own `header`.

Reply:
fixed